subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java

```diff
@@ -1003,8 +993,9 @@ private NodeInVisitingSegment(Node node, int visitingSegment) {
     }
 
     private static class NodeMapping extends AbstractCollection<Node> {
-        private final Map<Task, LocalTaskNode> taskMapping = Maps.newLinkedHashMap();
+        private final Map<WorkIdentity, Node> nodeMapping = Maps.newHashMap();
         private final Set<Node> nodes = Sets.newLinkedHashSet();
+        private final Set<Task> tasks = Sets.newLinkedHashSet();
```

Comment:
I don't understand why we need so many data structures here.

* `nodes` => `nodeMapping.values()`
* `tasks` => `nodeMapping.values().findAll { it instanceof TaskNode }.collect { it.task }`

Reply:
These are all easily solved, but we can follow up in a separate PR.